kernel/src/hil/flash.rs

```diff
@@ -142,3 +164,111 @@ pub trait Client<F: Flash> {
     /// Flash erase complete.
     fn erase_complete(&self, error: Error);
 }
+
+/// A checked flash region
+pub struct Page {
+    pub address: u64,
+    pub length: u32,
+}
+
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
+    /// `address` is calculated as an offset from the start of the flash
+    /// region.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    /// If `ErrorCode::NOSUPPORT` is returned then `read_page()`
+    // should be used instead.
+    fn read(
+        &self,
+        region: &Page,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Create a `Page` from the provided `address` and `length`.
+    ///
+    /// The `Page->address` and Page->length` will be modified to match
+    /// the flash hardware requirements.
+    fn get_read_region(&self, address: u64, length: u32) -> Result<Page, ErrorCode>;
+
+    /// Write data from a buffer to flash.
+    ///
+    /// This function will write the buffer `buf` to the `address` specified
+    /// in flash.
+    ///
+    /// `address` must be aligned to `W`.
+    /// The length of `buf` must be aligned to `W`.
+    ///
+    /// This function will not erase the page first. The user of this function
+    /// must ensure that a page is erased before writing.
+    /// Writes to flash can only turn a `1` to a `0`. To change a `0` to a `1`
+    /// the region must be erased.
+    ///
+    /// Note that some hardware only allows a limited number of writes before
+    /// an erase. If that is the case the implementation MUST return an error
+    /// `ErrorCode::NOMEM` when this happens, even if the hardware silently
+    /// ignores the write.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    fn write(
+        &self,
+        region: &Page,
```

Comment:
I'm not sure that Page is the right choice here, because it opens the possibility that the buffer is of different length than the page.